### PR TITLE
Add paypal-rest-sdk sample

### PR DESCRIPTION
**Pull request by Kevded**
Use Paypal-rest-sdk on firebase functions
[paypal-rest-sdks](https://developer.paypal.com/docs/api/rest-sdks/)

### CODE REVIEW

File: paypal/functions/index.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
+var functions = require('firebase-functions');
```
Comment (nicolasgarnier) [inline paypal/functions/index.js]: let's make it all ES2015 style code like the other samples. Can you replace the `var` with `const` ?

File: paypal/functions/index.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            },
+            // This is the payment transaction description. Maximum length: 127
+            description: 'Cloud Functions Firebase paypal-rest-sdk' // req.body.id
+            // reference_id string .Optional. The merchant-provided ID for the purchase unit. Maximum length: 256.
+            // reference_id: ''
+        }]
+    });
+
+    // Initialize the payment and redirect the user
+    paypal.payment.create(payReq, function (error, payment) {
```
Comment (nicolasgarnier) [inline paypal/functions/index.js]: You can use an arrow function here like:

```js
paypal.payment.create(payReq, (error, payment) => {
```

File: paypal/functions/index.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        var links = {};
+
+        if (error) {
+            console.error(JSON.stringify(error));
```
Comment (nicolasgarnier) [inline paypal/functions/index.js]: just:

```
console.error(error);
```

File: paypal/functions/index.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            res.end();
+        } else {
+            // Capture HATEOAS links
+            payment.links.forEach(function (linkObj) {
```
Comment (nicolasgarnier) [inline paypal/functions/index.js]: Use an arrow function here:

```js
payment.links.forEach(linkObj => {
```

File: paypal/functions/index.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            });
+
+            // If redirect url present, redirect user
+            if (links.hasOwnProperty('approval_url')) {
+                // REDIRECT USER TO links['approval_url'].href
+                // console.log(links['approval_url'])
+                res.set('Content-Type', 'application/json');
+                res.send({
+                    approval_url: links.approval_url.href
+                });
```
Comment (nicolasgarnier) [inline paypal/functions/index.js]: I think you can replace the 4 lines above by:

```js
res.json({approval_url: links.approval_url.href});
```

File: paypal/functions/index.js [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            } else {
+                console.error('no redirect URI present');
+            }
+            res.status('500').end();
```
Comment (nicolasgarnier) [inline paypal/functions/index.js]: That should go inside the `else` right ?

File: paypal/functions/index.js [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+});
+
+// Complete the payment
+exports.process = functions.https.onRequest((req, res) => {
+    var paymentId = req.query.paymentId;
+    var payerId = {
+        payer_id: req.query.PayerID
+    };
+
+    paypal.payment.execute(paymentId, payerId, function (error, payment) {
```
Comment (nicolasgarnier) [inline paypal/functions/index.js]: Use an arrow function here :)

File: paypal/functions/index.js [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        if (error) {
+            console.error(JSON.stringify(error));
```
Comment (nicolasgarnier) [inline paypal/functions/index.js]: just `console.error(error);`

File: paypal/functions/index.js [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+            res.redirect('http://localhost:4200/error');
```
Comment (nicolasgarnier) [inline paypal/functions/index.js]: Change that URL to a production URL ?

File: paypal/functions/index.js [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+                    'paid': true,
+                    'description': description,
+                    'date': Date.now()
+                }).then(r => console.info('promise: ', r));
+                // replace url
+                res.redirect('http://localhost:4200/success');
+            } else {
+                console.warn('payment.state: not approved ?');
+                // replace debug url
+                res.redirect('https://console.firebase.google.com/project/<project-id>/functions/logs?search=&severity=DEBUG');
```
Comment (nicolasgarnier) [inline paypal/functions/index.js]: You can replace the project ID automatically here as shown above.

File: paypal/functions/index.js [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+            } else {
+                console.error('no redirect URI present');
+            }
+            res.status('500').end();
+        }
+    });
+});
+
+// Complete the payment
+exports.process = functions.https.onRequest((req, res) => {
```
Comment (nicolasgarnier) [inline paypal/functions/index.js]: Could you add  a little bit more details in the comment here to explain a bit more what this function does?

File: paypal/functions/index.js (hunk 5 above)
Comment (nicolasgarnier) [inline paypal/functions/index.js]: Could we also do an HTTP Redirect (like a 302) ?

**Comment by nicolasgarnier** [comment]
ping :)

**Comment by Kevded** [comment]
I work on it. It just misses the comments. Sorry I was taken because of school.

Deddy SAINT-VAL



De : Nicolas Garnier
Envoyé : jeudi 29 juin 14:24
Objet : Re: [firebase/functions-samples] Add paypal-rest-sdk sample (#151)
À : firebase/functions-samples
Cc : kevded, Author


ping :)

**Comment by nicolasgarnier** [comment]
Thanks @Kevded  I've updated the README with more detailed step-by-step instructions.

Do you think it would be possible for you to also commit a small web UI like the one you showed in the animated GIF? this would make the sample so much more complete! PS if you do don't forget about this feature: https://firebase.google.com/docs/hosting/functions this will allow you to have the Paypal Cloud Function on the same domain as the web UI.

**Comment by Kevded** [comment]
I tried to pass the uid value via the optional "custom" field in the "transaction" object as in paypal [https://developer.paypal.com/docs/api/payments/#definition-transaction](https://developer.paypal.com/docs/api/payments/#definition-transaction), but this does not work, so I have passed through the "description" field.

**Comment by nicolasgarnier** [comment]
I added a few style fixes and we should be all set :)

**Pull request merged**